src/util/newconfig/ArrayView.cpp

```diff
@@ -0,0 +1,72 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of clio: https://github.com/XRPLF/clio
+    Copyright (c) 2024, the clio developers.
+
+    Permission to use, copy, modify, and distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include "util/newconfig/ArrayView.hpp"
+
+#include "util/Assert.hpp"
+#include "util/newconfig/Array.hpp"
+#include "util/newconfig/ConfigDefinition.hpp"
+#include "util/newconfig/ConfigValue.hpp"
+#include "util/newconfig/ObjectView.hpp"
+#include "util/newconfig/ValueView.hpp"
+
+#include <algorithm>
+#include <cstddef>
+#include <stdexcept>
+#include <string_view>
+
+namespace util::config {
+
+ArrayView::ArrayView(std::string_view prefix, ClioConfigDefinition const& configDef)
+    : prefix_{prefix}, clioConfig_{configDef}
+{
+    auto it = std::find_if(configDef.map_.begin(), configDef.map_.end(), [this](auto const& pair) {
+        return pair.first.starts_with(prefix_);
+    });
+    ASSERT(it != clioConfig_.map_.end(), "prefix does not exist in config definition.");
+    ASSERT(prefix_.contains(".[]"), "Not an array");
+}
+
+ValueView
+ArrayView::valueAt(std::size_t idx) const
```

Comment:
I feel the interface is a bit weird here. The valueAt will cause error when prefix_ is just prefix, but size() is fine.

Reply:
This is because if there are let's say 2 values for "dosguard.[].whitelist", if we do valueAt(2), it is going to error out because there isn't a 3rd value there. Check https://github.com/XRPLF/clio/pull/1544#discussion_r1680301679 first 😄